### PR TITLE
MAIL-4 mail templates

### PR DESCRIPTION
**Pull request by tbg-jasonfung**
- x3 new mailer templates, `base.html.twig`, `base.text.twig` and `donor-donation-success.html.twig`

### CODE REVIEW

**Comment by tbg-jasonfung** [comment]
@NoelLH Any idea how to show images in the template? The html tag `<img` doesn't seem to work and the docs refer embedding an image when setting the body, but the way we're generating the body comes from the request params.

**Comment by NoelLH** [comment]
> @NoelLH Any idea how to show images in the template? The html tag `<img` doesn't seem to work and the docs refer embedding an image when setting the body, but the way we're generating the body comes from the request params.

@tbg-jasonfung SwiftMailer multipart or `embed(...)` shorthand is what you want.

This has nothing to do with the request params which make the original API request. The consumer that sends the email doesn't even have access to a Request! Let's get on a call if this isn't clear.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 1]
```diff
@@ -56,27 +56,37 @@ public function __invoke(SendRequest $sendRequest): void
...
         }
 
-        $bodyRenderedHtml = $this->twig->render("{$sendRequest->templateKey}.html.twig", $sendRequest->params);
+        // Instantiate a new Swift Message object
+        $email = new \Swift_Message();
+
+        $pathToImages = __DIR__ . '/../../../../images/';
+
+        $images['header'] = $email->embed(\Swift_Image::fromPath($pathToImages. 'TBG.jpg'));
+        $images['footer'] = $email->embed(\Swift_Image::fromPath($pathToImages . 'CCh.png'));
```
Comment (tbg-jasonfung) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:65]: @NoelLH So I've got the embed image to work, however they're showing up as attachments in Gmail, and just going through the [docs ](https://swiftmailer.symfony.com/docs/messages.html) it mentions this:

`The embedded files are sent in the email as a special type of attachment that has a unique ID used to reference them within your HTML attributes. On mail clients that do not support embedded files they may appear as attachments.`

Not ideal :/ - I've tried replacing `$images['header']` with this: `$images['header'] = someImageURL` and it's showing as a broken image.

Any ideas?

Comment (NoelLH) [reply src/Application/Messenger/Handler/SendRequestConsumer.php:65]: @tbg-jasonfung Gmail supports embedded images. This is decades-old tech, and we're not doing the data URL thing they dislike discussed here. https://stackoverflow.com/a/42014708/2803757

I checked on some examples from when I last did this in 2016 and they load the image fine in Gmail. It also detects that part as having been embedded and doesn't show it as an attachment.

So, something is wrong with the CID refs or paths you're using, or possibly the types / dispositions of the parts in the mail. Have a look at the source and what the parts are named / encoded as?

Comment (tbg-jasonfung) [reply src/Application/Messenger/Handler/SendRequestConsumer.php:65]: > Not ideal :/ - I've tried replacing `$images['header']` with this: `$images['header'] = someImageURL` and it's showing as a broken image.

Actually I take that back, looks like using SF public links don't work, i'll try using S3

Comment (NoelLH) [reply src/Application/Messenger/Handler/SendRequestConsumer.php:65]: @tbg-jasonfung let's not get bogged down in different remote image solutions. We don't need any.

One thing I've just noticed is that in the other case I passed the HTML as the "primary" body in `setBody()`, which makes more sense as only the HTML can reference embedded images. Try swapping how the HTML and text parts get set up?

Comment (tbg-jasonfung) [reply src/Application/Messenger/Handler/SendRequestConsumer.php:65]: @NoelLH Yes! That did the trick! Thanks

**Comment by tbg-jasonfung** [comment]
image

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 2]
```diff
@@ -56,27 +56,36 @@ public function __invoke(SendRequest $sendRequest): void
             $this->fail($sendRequest->id, "Validation failed: {$this->validator->getReason()}");
         }
 
-        $bodyRenderedHtml = $this->twig->render("{$sendRequest->templateKey}.html.twig", $sendRequest->params);
+        // Instantiate a new Swift Message object
+        $email = new \Swift_Message();
+
+        $pathToImages = __DIR__ . '/../../../../images/';
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:62]: This is a bit hard to follow. You can give `dirname()` a number of directory hops now, so doing this relative to `__DIR__` is now common.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 3]
```diff
@@ -56,27 +56,36 @@ public function __invoke(SendRequest $sendRequest): void
...
+
+        $images['header'] = $email->embed(\Swift_Image::fromPath($pathToImages . 'TBG.jpg'));
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:64]: These get mushed in with the other params, so this naming is not very maintainable. How about `headerImageRef` etc. or an array key `imageRefs` with these as inner array elements?

As mentioned on our call I think this boilerplate would also be good to abstract out to an isolated helper method. At the least we should avoid doing all the same path gubbins as a copy/paste.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 4]
```diff
@@ -56,27 +56,36 @@ public function __invoke(SendRequest $sendRequest): void
...
+        $images['footer'] = $email->embed(\Swift_Image::fromPath($pathToImages . 'CCh.jpg'));
+
+        $params = array_merge($images, $sendRequest->params);
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:67]: Maybe call this `$templateMergeParams` or similar? It gets pretty hard to follow the different levels without good names.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 5]
```diff
@@ -56,27 +56,36 @@ public function __invoke(SendRequest $sendRequest): void
...
+
+        $bodyRenderedHtml = $this->twig->render("{$sendRequest->templateKey}.html.twig", $params);
         $bodyPlainText = strip_tags($bodyRenderedHtml);
 
         $config = $this->config->get($sendRequest->templateKey);
 
         // For each $p in the configured subjectParams, we need an array element with $emailData->params[$p].
-        $subjectMergeValues = array_map(fn($param) => $sendRequest->params[$param], $config->subjectParams);
+        $subjectMergeValues = array_map(fn($param) => $params[$param], $config->subjectParams);
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:75]: This change adds cognitive overhead we don't need: subject lines don't use images.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 6]
```diff
@@ -56,27 +56,36 @@ public function __invoke(SendRequest $sendRequest): void
...
-        $email = (new \Swift_Message())
-            ->addTo($sendRequest->recipientEmailAddress)
-            ->setSubject($subject)
-            ->setBody($bodyPlainText)
-            ->addPart($bodyRenderedHtml, 'text/html')
-            ->setContentType('text/html')
-            ->setCharset('utf-8')
-            ->setFrom(getenv('SENDER_ADDRESS'));
+        $email->addTo($sendRequest->recipientEmailAddress);
+        $email->setSubject($subject);
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:83]: We could still use chained fluent calls after the first `$email` – personally I find it more readable in cases like this where every line performs a clean operation on the same object.

File: templates/_base.html.twig [hunk 7]
```diff
@@ -2,18 +2,24 @@
...
-    }
-    </style>
   </head>
 <body>
-  <div id="content">{% block content %}{% endblock %}</div>
-  <div id="footer">
-    {% block footer %}
-        The Big Give xx
-    {% endblock %}
+  <div style="margin: 0 auto; max-width:500px;">
```
Comment (NoelLH) [inline templates/_base.html.twig:7]: How about a class or ID to keep the markup readable? The point of the CSS inlining plugin is to support this.

File: templates/_base.html.twig [hunk 8]
```diff
@@ -2,18 +2,24 @@
...
+    <img src="{{ header }}" alt="TheBigGive" title="The Big Give" width="250" height="50"/>
```
Comment (NoelLH) [inline templates/_base.html.twig:8]: If we're fixing the width at 250px we should serve the image at (at most) 500px for high-DPI displays. Also not sure what the reason is for scaling this way instead of with responsive CSS?

File: templates/_base.html.twig [hunk 9]
```diff
@@ -2,18 +2,24 @@
...
+    <div id="content" style="text-align: left;">
```
Comment (NoelLH) [inline templates/_base.html.twig:9]: As above

File: templates/_base.html.twig [hunk 10]
```diff
@@ -2,18 +2,24 @@
...
+    </div>
+    <div id="footer">
+      {% block footer %}
+        <p>
+          Kind regards,<br>
+          The Big Give Team<br>
+          www.thebiggive.org.uk<br>
+        </p>
+        <p>Powered by Charity Checkout</p>
+        <img src="{{ footer }}" alt="CharityCheckout" title="Charity Checkout" width="97" height="67"/>
```
Comment (NoelLH) [inline templates/_base.html.twig:20]: As above

File: templates/donor-donation-success.html.twig [hunk 11]
```diff
@@ -1,5 +1,40 @@
...
+    Reference on credit/debit card statement: <span>Charity</span>Checkout<span>.co</span>.uk
+  </p>
+
+  {% if tipAmount > 0 %}
+    <p>Thank you for leaving a tip of <strong>£{{ tipAmount }}</strong> toward the Big Give.</p>
+  {% endif %}
+
+  <p>Total amount you have been charged is <strong>£{{ totalChargedAmount }}</strong>.</p>
+
+  {# {% if comms %}
```
Comment (NoelLH) [inline templates/donor-donation-success.html.twig:28]: How's this going to work?

File: templates/donor-donation-success.html.twig [hunk 12]
```diff
@@ -1,5 +1,40 @@
...
+        <li>Donation: <strong>£{{ donationAmount}}</strong></li>
+        <li>Gift Aid: <strong>£{{ giftAidAmountClaimed }}</strong></li>
+        {% if matchedAmount > 0 %}
+            <li>Match Funds: <strong>£{{ matchedAmount }}</strong></li>
+        {% endif %}
+      </ul>
+
+  <p>
+    Donation reference: <strong>{{ transactionId }}</strong><br>
+    Reference on credit/debit card statement: <span>Charity</span>Checkout<span>.co</span>.uk
```
Comment (NoelLH) [inline templates/donor-donation-success.html.twig:19]: This doesn't feel great. How are screenreaders going to report all these separated bits?

I think it would be better to just leave it alone. If there's a concern that a link is distracting in terms of appearance, one solution would be to make it an actual, deliberate link but style it to look like normal text. It can/should still go to the domain so that donors who want to click it can.

File: templates/_base.html.twig (hunk 7 above)
Comment (tbg-jasonfung) [reply templates/_base.html.twig:7]: @NoelLH Just a quick one on this, I tried running this command `composer req twig/cssinliner-extra` following this [guide](https://twig.symfony.com/doc/2.x/filters/inline_css.html) but for some reason composer is complaining I don't have redis extension:

image

I've downloaded the windows redis binary but it's still not liking it, any ideas?

Comment (NoelLH) [reply templates/_base.html.twig:7]: @tbg-jasonfung you need to install the Redis PHP extension as it says, as well as a Redis client. Or just run this inside the Docker container, which already has both.

> @NoelLH Just a quick one on this, I tried running this command `composer req twig/cssinliner-extra` following this [guide](https://twig.symfony.com/doc/2.x/filters/inline_css.html) but for some reason composer is complaining I don't have redis extension:
> 
> image
> 
> I've downloaded the windows redis binary but it's still not liking it, any ideas?

Comment (NoelLH) [reply templates/_base.html.twig:7]: As discussed, this should also be unnecessary. The plugin detailed in the readme is already registered with SwiftMailer which should achieve the same thing without the extra template boilerplate.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 13]
```diff
@@ -56,9 +58,24 @@ public function __invoke(SendRequest $sendRequest): void
...
+
+        $images['headerImageRef'] = $this->embedImages($email, 'TBG.jpg');
+        $images['footerImageRef'] = $this->embedImages($email, 'CCh.jpg');
+
+        $templateMergeParams = array_merge($images, $sendRequest->params);
+
+        $bodyRenderedHtml = $this->twig->render("{$sendRequest->templateKey}.html.twig", $templateMergeParams);
         $bodyPlainText = strip_tags($bodyRenderedHtml);
 
+        $cssToInlineStyles->setUseInlineStylesBlock(false)
```
Comment (tbg-jasonfung) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:73]: @NoelLH Can you have a look and see whether i'm doing this right? I've followed the guidance in the readme `https://github.com/OpenBuildings/swiftmailer-css-inliner` but it's doesn't seem to be working as expected, i.e no emails are being sent at all

Comment (NoelLH) [reply src/Application/Messenger/Handler/SendRequestConsumer.php:73]: @tbg-jasonfung have you tried just leaving the config as it is? The plugin is already registered with SwiftMailer, and pretty sure it was working with my original example – I changed `p` size to check. No PHP changes needed.

If the email isn't sending, you'll have an error somewhere. Check the container logs to isolate the problem.

Comment (tbg-jasonfung) [reply src/Application/Messenger/Handler/SendRequestConsumer.php:73]: @NoelLH It does work, however it shows the styling tags in the preview: 
image

Comment (NoelLH) [reply src/Application/Messenger/Handler/SendRequestConsumer.php:73]: Not really enough info here @tbg-jasonfung, I would print out the HTML on the app side and look for the error. Also check if the plain text part tag stripping is working.

Comment (tbg-jasonfung) [reply src/Application/Messenger/Handler/SendRequestConsumer.php]: @NoelLH I've printed out the plain text part with tags removed, looks like it's kept the css classes, is there also a way to strip these out?

```
string(1434) "

      p, ul, li {
        font-size: 0.9rem;
      }

      blockquote {
        font-size: 0.7rem;
        font-style: italic;
      }

      .container {
        margin: 0 auto;
        max-width: 500px;
      }

      .content {
        text-align: left;
      }

      .headerImg {
        max-width: 50%;
        height: auto;
      }

      .footerImg {
        max-width: 30%;
        height: auto;
      }

        Dear Patti Smith,
  Thank you for supporting The charity with your generous gift through the Big Give.

    Here is your donation information:

        Total for The charity: £600.02
        Donation: £400.01
        Gift Aid: £100
                    Match Funds: £200.01

    Donation reference: d290f1ee-6c54-4b01-90e6-d701748f0851
    Reference on credit/debit card statement: CharityCheckout.co.uk

      Thank you for leaving a tip of £10 toward the Big Give.

  Total amount you have been charged is £410.01.

      A message from The charity:
    This is a really long message from the charity This is a really long message from the charity This is a really long message from the charity

  If you have any questions, please contact us.

          Kind regards,
          The Big Give Team
          www.thebiggive.org.uk

        Powered by Charity Checkout
```

Comment (NoelLH) [reply src/Application/Messenger/Handler/SendRequestConsumer.php]: Ah yes, makes sense @tbg-jasonfung if you read the docs https://www.php.net/manual/en/function.strip-tags.php I misremembered the scope of that fn.

Lots of ways around this. A regex to remove everything outside the content div after rendering might be quickest, though not very elegant or easy to maintain.

We could also think about conditionally extending a different base depending on the context, and render twice. We're going to discard everything outside the inner template except maybe the footer sign-off copy, so it's probably more maintainable to have a different base template which makes that explicit and lets us manage those parts independently if desired. We could pass in a `renderFormat` context var and conditionally extend either a `_base.html.twig` or `_base.text.twig`. Notes on that [here](https://ourcodeworld.com/articles/read/467/how-to-conditionally-extend-a-template-or-import-a-macro-in-twig).

Comment (tbg-jasonfung) [reply src/Application/Messenger/Handler/SendRequestConsumer.php]: @NoelLH I've literally just added a preg_replace + expression, I think this should achieve without the additional base template, let me know if you think otherwise

File: templates/donor-donation-success.html.twig (hunk 11 above)
Comment (tbg-jasonfung) [reply templates/donor-donation-success.html.twig:28]: Not quite sure tbh, I found it in the pardot email template and wondered if it's worth keeping

Comment (NoelLH) [reply templates/donor-donation-success.html.twig:28]: If we don't know why it's there it should not be going into new production code @tbg-jasonfung!

File: src/Application/Messenger/Handler/SendRequestConsumer.php (hunk 13 above)
Comment (NoelLH) [reply src/Application/Messenger/Handler/SendRequestConsumer.php]: @tbg-jasonfung I think a distinct base is more maintainable as explained above.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 14]
```diff
@@ -56,8 +56,23 @@ public function __invoke(SendRequest $sendRequest): void
...
+        $images['headerImageRef'] = $this->embedImages($email, 'TBG.png');
+        $images['footerImageRef'] = $this->embedImages($email, 'CCh.png');
+
+        $templateMergeParams = array_merge($images, $sendRequest->params);
+
+        $bodyRenderedHtml = $this->twig->render("{$sendRequest->templateKey}.html.twig", $templateMergeParams);
+
+        // We conduct a preg replace here because `strip_tags()` doesn't remove css class names between <style></style>
+        // Inspired by: https://stackoverflow.com/a/18089200/6620085
+        $bodyRenderedHtmlWithoutStyles = preg_replace("|<style\b[^>]*>(.*?)</style>|s", "", $bodyRenderedHtml);
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:71]: As per other comment, a base template is more flexible. This is also not foolproof, e.g. if we decide to use multiple style tags for the HTML, or other similar tags, it will break.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 15]
```diff
@@ -98,4 +112,10 @@ private function fail(string $id, string $reason)
         $this->logger->error("Sending ID $id failed: $reason");
         throw new RuntimeException($reason);
     }
+
+    private function embedImages($email, $fileName)
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:116]: Document? Return type? Strict argument types?

File: templates/_base.html.twig [hunk 16]
```diff
@@ -3,17 +3,52 @@
...
+    </div>
+    <div id="footer">
+      {% block footer %}
+        <p>
+          Kind regards,<br>
+          The Big Give Team<br>
+          www.thebiggive.org.uk<br>
+        </p>
+        <p>Powered by Charity Checkout</p>
+        <img class="footerImg" src="{{ footerImageRef }}" alt="CharityCheckout" title="Charity Checkout"/>
```
Comment (NoelLH) [inline templates/_base.html.twig:49]: Do we need `title`? Space in `alt` is probably more usable for screen readers.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 17]
```diff
@@ -98,4 +106,15 @@ private function fail(string $id, string $reason)
         $this->logger->error("Sending ID $id failed: $reason");
         throw new RuntimeException($reason);
     }
+
+    /**
+     * @param object $email
+     * @param string $fileName
+     * @return Swift_Image returns the embeded image
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:113]: @tbg-jasonfung in modern PHPs we can + should specify a real return type. This reveals errors, e.g. in this case we are returning a string, which make docs more accurate.

File: src/Application/Messenger/Handler/SendRequestConsumer.php [hunk 18]
```diff
@@ -56,8 +58,16 @@ public function __invoke(SendRequest $sendRequest): void
...
+        $email = new Swift_Message();
+
+        $additionalParams['headerImageRef'] = $this->embedImages($email, 'TBG.png');
+        $additionalParams['footerImageRef'] = $this->embedImages($email, 'CCh.png');
+        $additionalParams['renderHtml'] = true;
+
+        $templateMergeParams = array_merge($additionalParams, $sendRequest->params);
+
+        $bodyRenderedHtml = $this->twig->render("{$sendRequest->templateKey}.html.twig", $templateMergeParams);
+        $bodyPlainText = $this->twig->render("{$sendRequest->templateKey}.html.twig", $sendRequest->params);
```
Comment (NoelLH) [inline src/Application/Messenger/Handler/SendRequestConsumer.php:70]: @tbg-jasonfung  We still need to also `strip_tags` since the inner template is HTML.

I've added this and the spacing is weird but otherwise it works, so I think it's good enough for emails to have an accessible fallback that doesn't look too broken.
